beacon-chain/sync/data_columns_sampling.go

```diff
@@ -50,198 +52,225 @@ func sortedListFromMap(m map[uint64]bool) []uint64 {
+	for _, column := range columnsToSample {
+		remainingColumnsToSample[column] = true
 	}
 
 	// Get the active peers from the p2p service.
-	activePeers := s.cfg.p2p.Peers().Active()
+	activePids := s.cfg.p2p.Peers().Active()
 
-	var err error
+	// Query all peers until either all columns to request are retrieved or all active peers are queried (whichever comes first).
+	retrievedColumnsCount := 0
 
-	// Sampling is done sequentially peer by peer.
-	// TODO: Add parallelism if (probably) needed.
-	for _, pid := range activePeers {
-		// Early exit if all needed columns are already sampled. (This is the happy path.)
-		if len(missingColumns) == 0 {
-			break
+	for i := 0; len(remainingColumnsToSample) > 0 && i < len(activePids); i++ {
+		// Get the peer ID.
+		pid := activePids[i]
+
+		// Get the custody columns of the peer.
+		peerCustodyColumns, err := s.custodyColumnsFromPeer(pid)
+		if err != nil {
+			return 0, errors.Wrap(err, "custody columns from peer")
+		}
+
+		// Compute the intersection of the peer custody columns and the remaining columns to request.
+		peerRequestedColumns := make(map[uint64]bool, len(peerCustodyColumns))
+		for column := range remainingColumnsToSample {
+			if peerCustodyColumns[column] {
+				peerRequestedColumns[column] = true
+			}
+		}
+
+		// Remove the newsly requested columns from the remaining columns to request.
+		for column := range peerRequestedColumns {
+			delete(remainingColumnsToSample, column)
 		}
 
 		// Sample data columns from the peer.
-		missingColumns, err = s.sampleDataColumnFromPeer(pid, missingColumns, requestedRoot)
+		peerRetrievedColumns, err := s.sampleDataColumnsFromPeer(pid, peerRequestedColumns, root)
 		if err != nil {
-			return errors.Wrap(err, "sample data column from peer")
+			return 0, errors.Wrap(err, "sample data columns from peer")
 		}
+
+		// Update the retrieved columns.
+		retrievedColumnsCount += len(peerRetrievedColumns)
 	}
 
-	requestedColumnsList := sortedListFromMap(requestedColumns)
+	return uint64(retrievedColumnsCount), nil
+}
 
-	if len(missingColumns) == 0 {
-		log.WithField("requestedColumns", requestedColumnsList).Debug("Successfully sampled all requested columns")
-		return nil
-	}
+// incrementalDAS samples data columns from active peers using incremental DAS.
+// https://ethresear.ch/t/lossydas-lossy-incremental-and-diagonal-sampling-for-data-availability/18963#incrementaldas-dynamically-increase-the-sample-size-10
+func (s *Service) incrementalDAS(root [fieldparams.RootLength]byte, sampleCount uint64) error {
+	// Retrieve the number of columns.
+	columnsCount := params.BeaconConfig().NumberOfColumns
 
-	missingColumnsList := sortedListFromMap(missingColumns)
-	log.WithFields(logrus.Fields{
-		"requestedColumns": requestedColumnsList,
-		"missingColumns":   missingColumnsList,
-	}).Warning("Failed to sample some requested columns")
+	// Ramdomize all columns.
+	columns := randomSlice(columnsCount)
```

Comment:
we want to only sample columns that we do not custody, correct ? this function would do so for all our columns

Reply:
That's actually not 100% clear to me.

If we do so, nodes custodying all the columns won't sample any columns.
Node custodying at least 50% of the columns will reconstruct and save the 50% missing.
==> Node custodying 50% of the columns will actually custody 100% of the columns. Should they sample columns as well or not?

Reply:
It should definitely be distinct, there is 0 benefit in sampling a column you already know is available via custody. If you sample it again it doesn't push anything forward in whether we know a particular block is available.  
> If we do so, nodes custodying all the columns won't sample any columns.
Node custodying at least 50% of the columns will reconstruct and save the 50% missing.

If you see all the columns from gossip, then there isn't any benefit to sampling them, no ? The benefit I see is if you do not actually see this via gossip.

Reply:
New design:

- If we custody `>= 64` columns ==> No need to sample. Details are:
    - If we receive these columns, then we are able to reconstruct everything ==> No need to sample
    - If some are missing, then the block won't be eligible for the fork choice ==> No need to samle
- If we custody `< 64` columns:
    - We start to sample with incrementalDAS `min(16, 64-ourCustodyColumnsCount)`, as soon as we receive the block, columns we should NOT custody.

Reply:
Fixed in 92a1bfe80c8614092a631fd6e0da245c09b14c7a.